### PR TITLE
primitives

### PR DESCRIPTION
**Pull request by akellanotavailable**

### CODE REVIEW

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    //variable is in positive range
+    int testedVariable = 1;
+    //adding positive number
+    testedVariable += Integer.MAX_VALUE;
+
+    //for some reason variable is now negative(((
+    assertTrue(testedVariable < 0);
+
+  }
```
Comment (Alex99999999) [inline module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:21]: Add test to illustrate underflow

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Test
+  public void failCastTest() {
+    //long max value is far more than int max value
+    //this operation results in overflow
+    long extremelyBigLong = Long.MAX_VALUE;
+    assertNotEquals(extremelyBigLong, (int) extremelyBigLong);
+
```
Comment (Alex99999999) [inline module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:29]: What will happen if we cast a narrower integer type to a broader integer type?

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    //double stores the number the different way compared to float
+    float pi = 3.141592F;
+    assertNotEquals(Float.toString(pi), Double.toString(pi));
+
+  }
```
Comment (Alex99999999) [inline module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:34]: Add tests to illustrate upcasting and downcasting of floating point types without using wrapper classes.

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    assertNotEquals(0.9d, sneakyDouble);
+
+    double seackyDoubleBrother = 0.0;
+    for (int i = 1; i <= 10; i++) {
+      seackyDoubleBrother += 0.1;
+    }
+    assertNotEquals(1d, seackyDoubleBrother);
+
+  }
+
```
Comment (Alex99999999) [inline module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:49]: Why does it happen? 
Will it be the same result if add double to double and float to double?

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java (hunk 1 above)
Comment (akellanotavailable) [reply module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:21]: i guess that's overcoding, underflow is basically the same mechanism as overflow

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java (hunk 2 above)
Comment (akellanotavailable) [reply module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:29]: Casting narrower types to broader is considered a safe cast and will always be casted correctly. Implicit casting is performed internally by compiler, explicit is forced via relevant synaxis.  Explicit is used when the user is confident in the safety of cast and for some reason needs a different type.

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java (hunk 3 above)
Comment (akellanotavailable) [reply module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:34]: I don't think that upcasting and downcasting is a right term to use in this case. Also, i think that casting double to float without converting it to string would not be that illustrative due to compiler optimizations. I'll add them though.

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java (hunk 4 above)
Comment (akellanotavailable) [reply module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:57]: It happens due to specifics of the way floating numbers are stored in binary representation. That is relevant for both float and double, so the result will be exactly the same.

File: module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    assertTrue(testedVariable < 0);
+
+  }
+
+  @Test
+  public void failCastTest() {
+    //long max value is far more than int max value
+    //this operation results in overflow
+    long extremelyBigLong = Long.MAX_VALUE;
+    assertNotEquals(extremelyBigLong, (int) extremelyBigLong);
```
Comment (Provsky) [inline module1/src/test/java/net/broscorp/primitives/PrimitivesTest.java:28]: В двух словах опиши, почему числа не равны?

**Pull request closed without merging**